arch/cpu/nrf/sys/lpm.c

```diff
@@ -0,0 +1,293 @@
+{
+  uint8_t max_pm = LPM_MODE_MAX_SUPPORTED;
+
+  if(rtimer_arch_next()) {
+    *next_rtimer_set = true;
+
+    /* find out the timer of the next rtimer interrupt */
+    *next_rtimer = rtimer_arch_next();
+
+    if(RTIMER_CLOCK_LT(*next_rtimer, now + 2)) {
+      max_pm = MIN(max_pm, LPM_MODE_AWAKE);
+    } else if(RTIMER_CLOCK_LT(*next_rtimer, now + STANDBY_MIN_DURATION)) {
+      max_pm = MIN(max_pm, LPM_MODE_SLEEP);
+    }
+  } else {
+    *next_rtimer_set = false;
+  }
+
+  return max_pm;
+}
+/*---------------------------------------------------------------------------*/
+static uint8_t
+check_next_etimer(rtimer_clock_t now, rtimer_clock_t *next_etimer, bool *next_etimer_set)
+{
+  uint8_t max_pm = LPM_MODE_MAX_SUPPORTED;
+
+  *next_etimer_set = false;
+
+  /* Find out the time of the next etimer */
+  if (etimer_pending()) {
+    clock_time_t next = etimer_next_expiration_time();
+    *next_etimer_set = true;
+    *next_etimer = next * (RTIMER_SECOND / CLOCK_SECOND);
+    if (RTIMER_CLOCK_LT(*next_etimer, now + STANDBY_MIN_DURATION)) {
+      max_pm = MIN(max_pm, LPM_MODE_SLEEP);
+    }
+  }
+
+  return max_pm;
+}
+/*---------------------------------------------------------------------------*/
+static uint8_t
+setup_sleep_mode(void)
+{
+  uint8_t max_pm = LPM_MODE_MAX_SUPPORTED;
+  uint8_t pm;
+
+  rtimer_clock_t now;
+  rtimer_clock_t trigger_etimer = 0;
+  rtimer_clock_t next_rtimer = 0;
+  rtimer_clock_t next_etimer = 0;
+  bool next_rtimer_set = false;
+  bool next_etimer_set = false;
+
+  /* Check if any events fired before we turned interrupts off. If so, abort */
+  if(LPM_MODE_MAX_SUPPORTED == LPM_MODE_AWAKE || process_nevents()) {
+    return LPM_MODE_AWAKE;
+  }
+
+  now = RTIMER_NOW();
+  pm = check_next_rtimer(now, &next_rtimer, &next_rtimer_set);
+  if(pm < max_pm) {
+    max_pm = pm;
+  }
+  pm = check_next_etimer(now, &next_etimer, &next_etimer_set);
+  if(pm < max_pm) {
+    max_pm = pm;
+  }
+  trigger_etimer = RTIMER_CLOCK_MAX;
+  if(max_pm == LPM_MODE_SLEEP) {
+    /* In sleep mode, HFXO is powered up, therefore we do  not have
+     * to wakeup earlier to powering up the HFXO */
+    if(next_etimer_set) {
+      /* Schedule the next system wakeup due to etimer */
+      if(RTIMER_CLOCK_LT(next_etimer, now + MIN_SAFE_SCHEDULE)) {
+        /* Too soon in future, keep the system awake */
+      } else if(RTIMER_CLOCK_LT(now + MAX_SLEEP_TIME, next_etimer)) {
+        /* Too far in future, use MAX_SLEEP_TIME instead */
+        trigger_etimer = now + MAX_SLEEP_TIME;
+      } else {
+        trigger_etimer = next_etimer;
+      }
+    } else {
+      /* No etimers set. Since by default the CH1 RTC fires once every clock tick,
+       * need to explicitly schedule a wakeup in the future to save energy.
+       * But do not stay in this mode for too long, otherwise watchdog will be trigerred. */
+      trigger_etimer = now + MAX_SLEEP_TIME;
```

Comment:
When is `trigger_etimer` set to MAX_SLEEP_TIME, `!next_etimer_set || RTIMER_CLOCK_LT(now + MAX_SLEEP_TIME, next_etimer)`?

Reply:
for me, your code is  harder to read, dont you think?

Reply:
"If next etimer is not set, or next etimer is more than MAX_SLEEP_TIME into the future"?

I'm trying to understand the logic in this part of the function, I'm not sure it is equivalent to what I wrote.

Reply:
let me explain in some words:

1. soc_rtc.c sets in the rtc_handler the next wakeup time in 1/CLOCK_SECOND ms.  (rtimer is not relevant here, as rimter will be woken up exactly to the defined time)
2. setup_sleep_mode in lpm.c can overewrite this next wakeup time.
3. if the next rtimer and etimer is  > (now+STANDBY_MIN_DURATION) (10ms) -> we go into deep_sleep otherwise into sleep

Sleep: 
if an etimer is set, we set the next wakup time to the next  etimer (will be max 10ms) 
if no etimer is set (but eg. a rtimer within then next 10ms) we sleep for MAX_SLEEP_TIME

DeepSleep: 
- if we have an rtimer set, set the wakeup a bit before this rtimer, to fireup the HFXO
- if an etimer is set, and it is before the rtimer, set the next wakeup time to this etimer

makes this sense?